src/databricks/labs/ucx/source_code/python_libraries.py

```diff
@@ -1,136 +1,51 @@
-        dependency = Dependency(WrappingLoader(container), library)
-        return MaybeDependency(dependency, [])
+    def register_library(self, path_lookup: PathLookup, library: Path) -> list[DependencyProblem]:
+        compatibility = self._whitelist.distribution_compatibility(library.name)
+        if compatibility.known:
+            return compatibility.problems
+        return self._install_library(path_lookup, library)
 
-    def _temporary_virtual_environment(self, path_lookup: PathLookup) -> Path:
+    @cached_property
+    def _temporary_virtual_environment(self):
         # TODO: for `databricks labs ucx lint-local-code`, detect if we already have a virtual environment
         # and use that one. See Databricks CLI code for the labs command to see how to detect the virtual
         # environment. If we don't have a virtual environment, create a temporary one.
         # simulate notebook-scoped virtual environment
-        if len(self._lib_install_folder) == 0:
-            self._lib_install_folder = tempfile.mkdtemp(prefix="ucx-python-libs-")
-            path_lookup.append_path(Path(self._lib_install_folder))
-        return Path(self._lib_install_folder)
+        lib_install_folder = tempfile.mkdtemp(prefix='ucx-')
+        return Path(lib_install_folder)
 
-
-class DistInfoContainer(SourceContainer):
-    # TODO: remove this class, as we don't need it anymore
-
-    def __init__(self, file_loader: FileLoader, white_list: Whitelist, package: DistInfo):
-        self._file_loader = file_loader
-        self._white_list = white_list
-        self._package = package
-
-    def build_dependency_graph(self, parent: DependencyGraph) -> list[DependencyProblem]:
-        problems: list[DependencyProblem] = []
-        for module_path in self._package.module_paths:
-            maybe = parent.register_dependency(Dependency(self._file_loader, module_path))
-            if maybe.problems:
-                problems.extend(maybe.problems)
-        for library_name in self._package.library_names:
-            compatibility = self._white_list.distribution_compatibility(library_name)
-            if compatibility.known:
-                # TODO attach compatibility to dependency, see https://github.com/databrickslabs/ucx/issues/1382
-                problems.extend(compatibility.problems)
-                continue
-            more_problems = parent.register_library(library_name)
-            problems.extend(more_problems)
-        return problems
-
-    def __repr__(self):
-        return f"<DistInfoContainer {self._package}>"
+    def _install_library(self, path_lookup: PathLookup, library: Path) -> list[DependencyProblem]:
+        """Pip install library and augment path look-up to resolve the library at import"""
+        venv = self._temporary_virtual_environment
+        path_lookup.append_path(venv)
+        return_code, _, stderr = self._runner(f"pip install {library} -t {venv}")
```

Comment:
Maybe we want to some sort of cache around the `_install_library` to avoid installing the same library multiple times. This for example happens for the wheel we define at the top of every notebook that we deploy with ucx. Pip will not fail, but it will be slow(er) than skipping the `pip install`.

OR 

We add ucx wheel to the known distributions to skip it and assume not many people have a setup where the same library is installed in various places - like we do

Reply:
Caching has to be done on the level of pip, as we don't have the cycles to correctly implement the equivalent of all pip coordinates possible (git repos, zip files, version specifiers, editable installs, etc).

Adding ucx to known.json is to save time during testing, right?..